### ISSUE TITLE
Incremental graphs in ggplot2 

### ISSUE THREAD

**Issue by weiyangtham** [chirunconf/chirunconf19#24]
I am a fan of showing graphs [layer by layer in slideshows](https://evangelinereynolds.netlify.com/post/layered-presentation-of-graphics-take-2/). I find that my workflow ends up being something like

1. Make the final version of the graph that I want
2. Deconstruct the graph into the desired increments

This ends up involving a lot of copying and pasting. It's not the worst thing since ggplot syntax fits well with the layering idea, but I have found it to be enough work that I don't always do it. I've thought it would be nice to have some functions that can take a plot and show you just the parts you want. Roughly, 

```
p = ggplot(mtcars, aes(mpg, wt)) + geom_point()
show_x_axis(p) # only shows x-axis
show_one_point(p) # only show one point so you can explain it
```

**Comment by maurolepore** [comment]
Sounds cool! Have you thought about how to implement it? 
My first (and only idea) would be to edit the final call to the ggplot with the rlang package by @lionel- et al.

**Comment by wlandau** [comment]
Yeah, this sounds awesome! Such a natural consequence of `ggplot2`'s layering paradigm.

In fact, `ggplot2` objects already have a `layers` element.

``` r
library(ggplot2)
pl <- ggplot(mtcars) +
  geom_line(aes(x = wt, y = mpg)) +
  ggtitle("mpg vs wt") +
  geom_point(aes(x = wt, y = mpg, color = as.factor(cyl))) +
  ggtitle("mpg vs wt with line")

pl
```

![](https://i.imgur.com/obVHtWf.png)

``` r

pl$layers
#> [[1]]
#> mapping: x = ~wt, y = ~mpg 
#> geom_line: na.rm = FALSE
#> stat_identity: na.rm = FALSE
#> position_identity 
#> 
#> [[2]]
#> mapping: x = ~wt, y = ~mpg, colour = ~as.factor(cyl) 
#> geom_point: na.rm = FALSE
#> stat_identity: na.rm = FALSE
#> position_identity

# Remove a layer.
pl$layers <- pl$layers[-1]

pl
```

![](https://i.imgur.com/54QRW05.png)

<sup>Created on 2019-03-07 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

But that's not what we want because both frames have "with line" in the title. Hence, @maurolepore's idea of computing on the language. A couple more ideas:

1. Walk the [abstract syntax tree (AST)](http://adv-r.had.co.nz/Expressions.html#ast-funs), discover `+` signs, and make a new increment for each one.
2. Define increments manually, which would still require walking the AST to discover calls to `increment()`.

```r
pl <- ggplot(mtcars) +
  geom_line(aes(x = wt, y = mpg)) +
  ggtitle("mpg vs wt") +
  increment() +
  geom_point(aes(x = wt, y = mpg, color = as.factor(cyl))) +
  ggtitle("mpg vs wt with line") +
  increment() 
```
